src/main/java/com/hms_networks/sc/cumulocity/data/CConnectorJsonDataPayload.java

```diff
@@ -0,0 +1,528 @@
+  }
+
+  /**
+   * Gets the {@link JSONObject} representing the data payload. This is the JSON object
+   * representation of the payload that can be sent to Cumulocity.
+   *
+   * @return the {@link JSONObject} representing the data payload
+   * @throws JSONException if an error occurs while building the JSON object
+   * @throws Exception if an error occurs while formatting the timestamp
+   * @since 1.0.0
+   */
+  public synchronized JSONObject getJsonObject() throws Exception {
+    // Create new JSONObject
+    JSONObject jsonObject = new JSONObject();
+
+    // Add time and type
+    jsonObject.put(KEY_TIME, SCTimeUtils.getIso8601FormattedTimestampForDate(time));
+    jsonObject.put(KEY_TYPE, type);
+
+    // Add external source if not null
+    if (externalSource != null) {
+      jsonObject.put(KEY_EXTERNAL_SOURCE, externalSource);
+    }
+
+    // Get aggregation method from config
+    CConnectorAggregationMethod queueDataAggregationMethod =
+        CConnectorMain.getConnectorConfig().getQueueDataAggregationMethod();
+
+    // Add fragments
+    Iterator fragmentsIterator =
+        fragments.entrySet().iterator(); // Iterator<Map.Entry<String,Fragment>>
+    while (fragmentsIterator.hasNext()) {
+      // Get fragment
+      Map.Entry fragmentEntry = (Map.Entry) fragmentsIterator.next(); // Map.Entry<String, Fragment>
+      String fragmentName = (String) fragmentEntry.getKey();
+      Fragment fragment = (Fragment) fragmentEntry.getValue();
+
+      // Create new fragment JSONObject
+      JSONObject fragmentJsonObject = new JSONObject();
+
+      // Add series
+      Iterator seriesIterator =
+          fragment.getSeriesMap().entrySet().iterator(); // Iterator<Map.Entry<String,List<Series>>>
+      while (seriesIterator.hasNext()) {
+        // Get series
+        Map.Entry seriesEntry =
+            (Map.Entry) seriesIterator.next(); // Map.Entry<String, List<Series>>
+        String seriesName = (String) seriesEntry.getKey();
+        List seriesList = (List) seriesEntry.getValue(); // List<Series>
+
+        // Create series list iterator
+        Iterator seriesListIterator = seriesList.iterator();
+
+        // Create series min/max/first/last/avg objects
+        Fragment.Series seriesMin = (Fragment.Series) seriesList.get(0);
+        Fragment.Series seriesMax = (Fragment.Series) seriesList.get(0);
+        Fragment.Series seriesFirst = (Fragment.Series) seriesList.get(0);
+        Fragment.Series seriesLast = (Fragment.Series) seriesList.get(0);
+        Object[] seriesValues = new Object[seriesList.size()];
+        int seriesValuesIndex = 0;
+
+        // Loop through seriesList and find min/max/first/last/avg
+        while (seriesListIterator.hasNext()) {
+          Fragment.Series seriesListItem = (Fragment.Series) seriesListIterator.next();
+
+          // Check min
+          Object currentSeriesValue = seriesListItem.getValue();
```

Comment:
Is there a way to optimize this so that we are calculating min only when min is the method?  It appears that we are performing unnecessary calculations here .

Reply:
Will optimize

Reply:
Optimized